modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_2_4/UpgradeDDMContent.java

```diff
@@ -0,0 +1,207 @@
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author Rodrigo Paulino
+ */
+public class UpgradeDDMContent extends UpgradeProcess {
+
+	public UpgradeDDMContent(JSONFactory jsonFactory) {
+		_jsonFactory = jsonFactory;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement ps1 = connection.prepareStatement(
+				StringBundler.concat(
+					"select contentId, data_, DDMStructureVersion.definition ",
+					"from DDMContent join DDMFormInstanceRecordVersion on  ",
+					"storageId = contentId join DDMFormInstanceVersion on ",
+					"DDMFormInstanceVersion.formInstanceId = ",
+					"DDMFormInstanceRecordVersion.formInstanceId and ",
+					"DDMFormInstanceVersion.version = ",
+					"DDMFormInstanceRecordVersion.formInstanceVersion join ",
+					"DDMStructureVersion on DDMStructureVersion.",
+					"structureVersionId = DDMFormInstanceVersion.",
+					"structureVersionId"));
+			PreparedStatement ps2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update DDMContent set data_ = ? where contentId = ?")) {
+
+			try (ResultSet rs = ps1.executeQuery()) {
+				while (rs.next()) {
+					String definition = rs.getString("definition");
+
+					JSONObject definitionJSONObject =
+						_jsonFactory.createJSONObject(definition);
+
+					String data = rs.getString("data_");
+
+					JSONObject dataJSONObject = _jsonFactory.createJSONObject(
+						data);
+
+					if (_upgradeDDMContentData(
+							dataJSONObject.getJSONArray("fieldValues"),
+							definitionJSONObject.getJSONArray("fields"))) {
+
+						ps2.setString(1, dataJSONObject.toJSONString());
+
+						long contentId = rs.getLong("contentId");
+
+						ps2.setLong(2, contentId);
+
+						ps2.addBatch();
+					}
+				}
+			}
+
+			ps2.executeBatch();
+		}
+	}
+
+	private DecimalFormat _getDecimalFormat(Locale locale) {
+		DecimalFormat decimalFormat = _decimalFormatsMap.get(locale);
+
+		if (decimalFormat == null) {
+			decimalFormat = (DecimalFormat)DecimalFormat.getInstance(locale);
+
+			decimalFormat.setGroupingUsed(false);
+			decimalFormat.setMaximumFractionDigits(Integer.MAX_VALUE);
+			decimalFormat.setParseBigDecimal(true);
+
+			_decimalFormatsMap.put(locale, decimalFormat);
+		}
+
+		return decimalFormat;
+	}
+
+	private HashMap<String, JSONObject> _mapDataFieldValues(
+		JSONArray fieldValuesJSONArray) {
+
+		HashMap<String, JSONObject> dataFieldValuesMap = new HashMap<>();
+
+		fieldValuesJSONArray.forEach(
+			object -> {
+				JSONObject fieldValueJSONObject = (JSONObject)object;
+
+				dataFieldValuesMap.put(
+					fieldValueJSONObject.getString("name"),
+					fieldValueJSONObject.getJSONObject("value"));
+			});
+
+		return dataFieldValuesMap;
+	}
+
+	private boolean _upgradeDDMContentData(
+		JSONArray fieldValuesJSONArray, JSONArray fieldsJSONArray) {
+
+		AtomicBoolean upgraded = new AtomicBoolean(false);
```

Comment:
Because I can't assign value to a variable inside a lambda expression. The `AtomicBoolean` works there just as a getter and setter for the boolean flag.

Reply:
Terrific! 😄